article: Efficient gas calculation algorithm for EVM (#123)
ethereum/evmone: docs/efficient_gas_calculation_algorithm.md

# Efficient gas calculation algorithm for EVM

This article describes how to efficiently calculate gas and check stack requirements
for Ethereum Virtual Machine (EVM) instructions.

## Instructions metadata

Let's start by defining some basic and universal instructions' parameters.

1. Base gas cost.
   
   This is the static gas cost of instructions. Some instructions may have 

>    additional cost depending on their operand values - these have to be

Reviewer: What about those whose gas costs also depend on the state?

Author: The same, you only include the minimal possible cost here. E.g. for `SSTORE` the base cost is 0 (old HFs) or 5000. And later might go to 200 with EIP-2200.

Reviewer: Can we assume the base gas cost to never be higher than the final gas cost?

Reviewer 2: yes

Reviewer: Can you add that to the text?

Author: > Can we assume the base gas cost to never be higher than the final gas cost?

In some way this is the definition of it.

Reviewer: I'm just saying that 'additional cost' could be interpreted as 'cost that has to be added' in which case you could also add a negative value. Furthermore, the sentence only explicitly states "depending on their operand values". I think it would be clearer if you also state "on their operand values or the environment", in which case there is no doubt left. Just my two cents :)

Author: Added both suggestions in the best way I was capable of.

   handled individually during the instruction execution.
   
2. Stack height requirement.

   This is the minimum stack height (number of items on the stack) 

>    required for the instruction execution.

Reviewer: ```suggestion
   required prior to the instruction execution.
```

   
3. Stack height change.

   This is difference of the stack height before and after the instruction 
   execution. Can be negative if the instruction pops more items than pushes.
   
Examples:

| opcode  | base gas cost | stack height requirement | stack height change |
| ------- | ------------- | ------------------------ | ------------------- |
| ADD     | 3             | 2                        | -1                  |
| EXP     | 50            | 2                        | -1                  |
| DUP4    | 3             | 4                        | 1                   |
| SWAP1   | 3             | 2                        | 0                   |
| ADDRESS | 2             | 0                        | 1                   |
| CALL    | 700           | 7                        | -6                  |

## Basic blocks

A _basic block_ is a sequence of instructions that are executed "straight-line"

> without being interrupted by jumps. I.e. they are nodes in the _control flow graph_.

Reviewer: What about calls and e.g. stop?

Reviewer: Ah stop is mentioned below, but it would be interesting to know about call, since a call does not really influence the internal control flow.

Author: As mentioned in Misc, for calls we only need to know the "current" gas counter, which is not directly available in EVMs following this algorithm. Because of that, in evmone we add back the cost of all instructions after the call.

> The name "basic block" has been taken from LLVM. Other names are: just "block"

Reviewer: Isn't this just the generic term in compiler theory?

Author: I never had proper education in this field. If the term is well known, I should simplify this paragraph.

Reviewer: Not sure about the well-known-ness, but llvm certainly did not introduce it :)

Reviewer 2: I'd prefer if you called it 'basic segment', and talk about segments. Because we're talking about blockchains here, and in extension maybe doing paralell gas-calculation of future blocks, and things become a bit messy if we have two definitions for 'block'.

Author: This is pretty concrete term: https://en.wikipedia.org/wiki/Basic_block.

Maybe "basic instruction block" then?

in wasm or "subroutine" in subroutine-threaded interpreters.

In EVM there are simple rules to identify basic block boundaries:

> 1. The following instructions _end_ a basic block:

Reviewer: ```suggestion
1. A basic block _ends_ after the following instructions:
```

   - `JUMP`,
   - `JUMPI`,
   - `STOP`,
   - `RETURN`,
   - `REVERT`,
   - `SELFDESTRUCT`,

>    - an instruction directly preceding `JUMPDEST`.

Reviewer: ```suggestion
   - the last instruction in the program
   - an invalid instruction
```

Author: We actually ignore invalid instructions by assuming they cost 0.

> 2. The following instructions _start_ a basic block:

Reviewer: ```suggestion
2. A basic block _starts_ right before the following instructions:
```

>    - `JUMPDEST`,

Reviewer: ```suggestion
```
(this should be redundant together with the two rules below)

Author: I don't believe this is true. You can have a `JUMPDEST` anywhere.

Reviewer: sure, but you say: a basic block starts right before ... an instruction directly after an intruction that has ended the previous basic block. And above: a basic block ends after ... an instruction directly preceding "jumpdest". Maybe just nitpicking :)

Reviewer 2: I'm wondering about the attackability of this -- in hte case where a contract consists of 24k `JUMPDEST`s. I guess that's up to the implementation.

Author: I have a test/benchmark for this one. Other variations about it are definitely needed.

evmone's top speed is ~1Ggas/s. In this case where we do gas/stack checking for every instruction the execution is ~0.5Ggas/s.

The bigger problem is the analysis of this code takes ~0.1 ms what means it should cost ~1000 gas. This is the biggest concern so far.

   - the first instruction in the code,
   - an instruction directly after an instruction that has ended previous basic block.

> 

Reviewer: ```suggestion
A basic block is a shortest sequence of instructions such that a basic block starts before the first instruction and ends after the last.
```

Reviewer: (hopefully, this is still well-defined...)

## Algorithm

The algorithm for calculating gas and checking stack requirements precomputes

> the values for basic blocks and during execution the checks are done once per block.

Reviewer: ```suggestion
the values for basic blocks and during execution the checks are done only once per block.
```

### Collecting requirements for basic blocks

> For a basic block we need to collect following information:

Reviewer: ```suggestion
For a basic block we need to collect the following information:
```

- total base gas required by instructions,
- the stack height required (the minimum stack height needed to execute 
  all instructions in the block),

> - ~~the start-to-end stack height change~~,

Reviewer: Why is this crossed out?

Author: When writing the article, I realized we don't have to keep this value for execution. It is only needed temporarily to calculate other values. But I want to apply this change to evmone first to make sure. I believe @zac-williamson proposed this change in #11.

Author: Change in evmone: https://github.com/ethereum/evmone/pull/125. So this is not needed.

> - the relative maximum stack height.

Reviewer: What exactly is this?

Author: The maximum stack height increase in the basic block relatively to the stack height on which the block execution starts. If `stack_max` is 5 and the block execution starts at stack height 10 the absolute stack height will reach 15.

Reviewer: So maybe: "The maximum stack height relative to the stack height at block start"? Also is that per instruction in the block or is it a single value?

Author: Per block.

Reviewer 3: Maybe "the maximum stack height change over the block"?
I would also call it `stack_change_max` in the pseudocode

Author: @axic proposed `stack_max_growth`, and I renamed is like that in the code.

    stack_change = 0  # FIXME: We don't have too keep it.
    stack_max = 0

def collect_basic_block_requirements(basic_block):
    for instruction in basic_block:
        basic_block.gas_required += instruction.gas_required
        
        current_stack_required = instruction.stack_required - basic_block.stack_change
        basic_block.stack_required = max(basic_block.stack_required, current_stack_required)
        
        basic_block.stack_change += instruction.stack_change
        
        basic_block.stack_max = max(basic_block.stack_max, basic_block.stack_change)
```

### Checking basic block requirements

During execution, before executing an instruction that starts a basic block,
the basic block requirements must be checked.

```python
class ExecutionState:
    gas_left = 0
    stack = []

def check_basic_block_requirements(state, basic_block):
    state.gas_left -= basic_block.gas_required
    if state.gas_left < 0:
        raise OutOfGas()
    
    if len(state.stack) < basic_block.stack_required:
        raise StackUnderflow()
    
    if len(state.stack) + basic_block.stack_max > 1024:
        raise StackOverflow()
```

## Misc

### EVM may terminate earlier

> Because requirements for a whole basic blocks are checked up front, the instructions

Reviewer: ```suggestion
Because requirements for a whole basic block are checked up front, the instructions
```

that have observable external effects might not be executed although they would be
executed if the gas counting would have been done per instruction.
This is not a consensus issue because the execution terminates with a "hard" exception
anyway (and all effects are reverted) but might produce unexpected traces 
or terminate with a different exception type.

### Current "gas left" value

> In EVMJIT also `GAS` and _call_ instructions begin a basic block. This is because

Reviewer 3: just a rewording
```suggestion
In EVMJIT additional instructions that begin a basic block are `GAS` and any of the _call_ instructions. This is because
```
